### DISCUSSION TITLE
Changing data in postfields

### DISCUSSION THREAD

**Discussion by Kolhun** [curl/curl#12882] [Q&A] [answered]
Hello everyone, a question arose, during the execution of my program, the string (the request body) changes and, accordingly, when trying to update the data for the curl query (when the string itself is updated), it simply does not do this, but uses the original data. Below is an example code of how it works. Thanks for the help
```cpp
void CurlMethod(Curl* _curl, std::string _dataPostString)
{
  CURLcode res;
  if (_curl != NULL)
  {
      curl_easy_setopt(_curl, CURLOPT_TIMEOUT, 100);
      curl_easy_setopt(_curl, CURLOPT_URL, (const char*)_data_url.c_str());
      curl_easy_setopt(_curl, CURLOPT_CUSTOMREQUEST, "POST");
      curl_easy_setopt(_curl, CURLOPT_WRITEFUNCTION, WriteCallback);
      curl_easy_setopt(_curl, CURLOPT_WRITEDATA, &_readBuffer);
      curl_easy_setopt(_curl, CURLOPT_ACCEPT_ENCODING, "utf-8");
      curl_easy_setopt(_curl, CURLOPT_HTTPHEADER, _headers);
      curl_easy_setopt(_curl, CURLOPT_POSTFIELDS, _dataPostString.c_str()); // <- THIS Starting value
      curl_easy_setopt(_curl, CURLOPT_VERBOSE, 1L);
  
      res = curl_easy_perform(_curl);
  
      if (res != CURLE_OK)
      {
          printf("\n# curl_easy_perform() failed: %s\n", curl_easy_strerror(res));        
      }
      else {
          printf("\n# curl_easy_perform() compleet\n");
      }
  
      curl_easy_cleanup(_curl);
  
  }
}

class main()
{
  Curl* curl = curl_easy_init();;
  CurlMethod( curl, _dataPostString.c_str());
  if(ButonClick(""))
  {
    //The contents of the _data Post String change here (And it really changes)
    ...
    //I'm trying to update the data for the request
    curl_easy_setopt( _curl, CURLOPT_POSTFIELDS, __dataPostString.c_str()); 
  }
}

```
The above code does not change postfields at all, what could be the reason?

**Comment by jzakrzewski** [comment, marked as answer]
I don't understand what exactly you're trying to achieve, but in any case this `curl_easy_setopt( _curl, CURLOPT_POSTFIELDS, __dataPostString.c_str());` is executed after the request has been performed, and the handle destroyed!

**Reply by Kolhun** [reply]
Does the above option not fulfill this condition? If not, tell me an example

**Reply by jay** [reply]
It's unclear how your code flows. CurlMethod sets the POST content and makes the request. In the lines before CURLOPT_POSTFIELDS is set add a line `fprintf(stderr, "setting POST data to %s\n", _dataPostString.c_str());`

**Reply by Kolhun** [reply]
When you first start curl
![image](https://github.com/curl/curl/assets/36815965/bfaa0686-6144-4979-b070-9bcd72120dba)
Here, the "limit" is 4, when changing the line, it also returns the same line (it should be 5 instead of 4)
![image](https://github.com/curl/curl/assets/36815965/5e970f8e-37dc-4cce-80c9-9655fd9f229f)

**Reply by jay** [reply]
It's hard to tell from what you've shown us but possibly c_str is invalid because the underlying string object is modified/destroyed before curl uses it.

See [CURLOPT_POSTFIELDS](https://curl.se/libcurl/c/CURLOPT_POSTFIELDS.html):

_The data pointed to is NOT copied by the library: as a consequence, it must be preserved by the calling application until the associated transfer finishes. This behavior can be changed (so libcurl does copy the data) by instead using the [CURLOPT_COPYPOSTFIELDS](https://curl.se/libcurl/c/CURLOPT_COPYPOSTFIELDS.html) option._

**Reply by Kolhun** [reply]
Yes, it really helped. Thank you so much for your help
